metricbeat/beater/metricbeat.go

```diff
@@ -20,72 +23,96 @@ Each event created has the following format:
 		"@timestamp": timestamp
 	}
 
-All documents are currently stored in one index called metricbeat. It is important to use an independent namespace
-for each MetricSet to prevent type conflicts. Also all values are stored under the same type "metricsets".
-
+All documents are stored in one index called metricbeat. It is important to use
+an independent namespace for each MetricSet to prevent type conflicts. Also all
+values are stored under the same type "metricsets".
 */
 package beater
 
 import (
-	"fmt"
+	"sync"
 
 	"github.com/elastic/beats/libbeat/beat"
 	"github.com/elastic/beats/libbeat/logp"
-	"github.com/elastic/beats/metricbeat/helper"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/pkg/errors"
 )
 
+// Metricbeat implements the Beater interface for metricbeat.
 type Metricbeat struct {
-	done   chan struct{}
-	config *Config
+	done    chan struct{}    // Channel used to initiate shutdown.
+	config  *Config          // Metricbeat specific configuration data.
+	modules []*moduleWrapper // Active list of modules.
 }
 
 // New creates and returns a new Metricbeat instance.
 func New() *Metricbeat {
 	return &Metricbeat{}
 }
 
-func (mb *Metricbeat) Config(b *beat.Beat) error {
+// Config unpacks the Metricbeat specific configuration data.
+func (m *Metricbeat) Config(b *beat.Beat) error {
 	// List all registered modules and metricsets.
-	logp.Info("%s", helper.Registry.String())
+	logp.Info("%s", mb.Registry.String())
 
-	mb.config = &Config{}
-	err := b.RawConfig.Unpack(mb.config)
+	m.config = &Config{}
+	err := b.RawConfig.Unpack(m.config)
 	if err != nil {
-		return fmt.Errorf("error reading configuration file. %v", err)
+		return errors.Wrap(err, "error reading configuration file")
 	}
 
 	return nil
 }
 
-func (mb *Metricbeat) Setup(b *beat.Beat) error {
-	mb.done = make(chan struct{})
-	return nil
+// Setup initializes the Modules and MetricSets that are defined in the
+// Metricbeat configuration.
+func (m *Metricbeat) Setup(b *beat.Beat) error {
+	var err error
+	m.done = make(chan struct{})
+	m.modules, err = newModuleWrappers(m.config.Modules, mb.Registry, b.Publisher)
+	return err
 }
 
-func (mb *Metricbeat) Run(b *beat.Beat) error {
-	// Checks all defined metricsets and starts a module for each entry with the defined metricsets
-	for _, moduleConfig := range mb.config.Metricbeat.Modules {
-
-		module, err := helper.Registry.GetModule(moduleConfig)
-		if err != nil {
-			return err
-		}
+// Run starts the workers for Metricbeat and blocks until Stop is called
+// and the workers complete. Each host associated with a MetricSet is given its
+// own goroutine for fetching data. The ensures that each host is isolated so
+// that a single unresponsive host cannot inadvertently block other hosts
+// within the same Module and MetricSet from collection.
+func (m *Metricbeat) Run(b *beat.Beat) error {
+	var wg sync.WaitGroup
+	for _, mw := range m.modules {
+		for _, msw := range mw.children {
```

Comment:
Is there a specific reason you move this back to the beater? I moved this previously to the module as I was thinking the beater should not have to deal with the complexity of Starting modules. So no knowledge about Hosts for example would be needed.